### PR TITLE
A4A: Add Sidebar navigation component

### PR DESCRIPTION
**Pull request by jkguidaven**
This pull request adds a new sidebar component for the A4A portal. The component is based on the Jetpack cloud and has been modified to fit the A4A use cases. It's worth noting that the background color will be updated once the brand color is added to the Color Studio.

Screenshot 2024-02-15 at 7 40 33 PM
Screenshot 2024-02-15 at 7 40 44 PM

Closes https://github.com/Automattic/jetpack-genesis/issues/222

## Proposed Changes

* Implement an A4A Sidebar component based on the Jetpack cloud sidebar. Most of the codes are copied from the source. 
* Add sample pages to demonstrate the new sidebar.

## Testing Instructions

* Checkout to this branch.
* Spin up the environment using the command yarn `start-a8c-for-agencies`.
* Goto http://agencies.localhost:3000/
* Confirm that the sidebar shows the logo and user menu in the footer.
* Click the **Sites** menu item to go to the Sites page
* Confirm that the Site selector and user menu are visible in the header and the 'Get help' link is visible in the footer.

**NOTE: Note that I didn't register /overview and /sites URL so please always start at `/`.**

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/a8c-for-agencies/controller.tsx [hunk 1]
```diff
@@ -1,9 +1,75 @@
-import { type Callback } from '@automattic/calypso-router';
+import page, { type Callback } from '@automattic/calypso-router';
+import { category, chevronLeft, home } from '@wordpress/icons';
+import Sidebar from './components/sidebar';
 
-export const a8cForAgenciesContext: Callback = ( context, next ) => {
-	context.header = <div>Header</div>;
-	context.secondary = <div>Secondary</div>;
-	context.primary = <div>Hello, World!</div>;
+export const redirectToOverviewContext: Callback = () => {
```
Comment (jkguidaven) [inline client/a8c-for-agencies/controller.tsx:5]: When rendering the UI from the root directory `/`, the master bar is displayed. In order to hide it, I need to redirect the page to `/overview`. We will address the master bar separately in a future pull request. This is a temporary redirect.

File: client/a8c-for-agencies/controller.tsx [hunk 2]
```diff
@@ -1,9 +1,75 @@
...
+	page( '/overview' );
+};
+
+export const overviewContext: Callback = ( context, next ) => {
```
Comment (jkguidaven) [inline client/a8c-for-agencies/controller.tsx:9]: Sample context that demonstrate sidebar format.

**Review by coder-karen — APPROVED** [review]
Works well based on the testing instructions, looks good to me for the initial skeleton of the navigation.

Also, you can see the implementation of the Jetpack Manage sites submenu here now: https://github.com/Automattic/wp-calypso/pull/87456.
I believe your implementation may end up being a bit different though as you will have an actual /sites/ route if I understand correctly.

**Comment by jkguidaven** [comment]
> Works well based on the testing instructions, looks good to me for the initial skeleton of the navigation.
> 
> Also, you can see the implementation of the Jetpack Manage sites submenu here now: #87456. I believe your implementation may end up being a bit different though as you will have an actual /sites/ route if I understand correctly.

Thank you for your review. I have seen your PR, and I will most likely have a similar implementation in the A4A portal. I am wondering if you have already explored ways to integrate the badge counter?

 Currently, this feature is not possible in the A4A portal until we have the Sites page. However, I am planning to implement it in a similar way to how it will be implemented in Jetpack Manage.

Screenshot 2024-02-15 at 4 51 52 PM

**Comment by coder-karen** [comment]
>  I am wondering if you have already explored ways to integrate the badge counter?

We've decided to leave that for now until the API allows for it... or do an API call to get all the sites and then go through all the sites to find the needs attention items, but that was deemed out of scope just now (for that PR and possibly project). But we can revisit if needed.

File: client/a8c-for-agencies/controller.tsx [hunk 3]
```diff
@@ -1,9 +1,75 @@
...
+};
+
+export const sitesContext: Callback = ( context, next ) => {
+	context.secondary = (
+		<Sidebar
+			path="/"
+			title="Sites"
+			description="sample text"
+			backButtonProps={ {
+				label: 'Back to overview',
```
Comment (yashwin) [inline client/a8c-for-agencies/controller.tsx:50]: Can we add translated text here?

File: client/a8c-for-agencies/controller.tsx [hunk 4]
```diff
@@ -1,9 +1,75 @@
...
+				onClick: () => {
+					page( '/overview' );
+				},
+			} }
+			menuItems={ [
+				{
+					icon: category,
+					path: '/',
+					link: '/sites',
+					title: 'Needs attention',
```
Comment (yashwin) [inline client/a8c-for-agencies/controller.tsx:61]: Can we add translated text here?

File: client/a8c-for-agencies/controller.tsx [hunk 5]
```diff
@@ -1,9 +1,75 @@
...
+	context.primary = <div>Overview page</div>;
+
+	next();
+};
+
+export const sitesContext: Callback = ( context, next ) => {
+	context.secondary = (
+		<Sidebar
+			path="/"
+			title="Sites"
```
Comment (yashwin) [inline client/a8c-for-agencies/controller.tsx:47]: Can we add translated text here?

File: client/a8c-for-agencies/controller.tsx [hunk 6]
```diff
@@ -1,9 +1,75 @@
...
+export const overviewContext: Callback = ( context, next ) => {
+	context.secondary = (
+		<Sidebar
+			path="/"
+			menuItems={ [
+				{
+					icon: home,
+					path: '/',
+					link: '/overview',
+					title: 'Overview',
```
Comment (yashwin) [inline client/a8c-for-agencies/controller.tsx:18]: Can we add translated text here?

File: client/a8c-for-agencies/controller.tsx [hunk 7]
```diff
@@ -1,9 +1,75 @@
...
+					trackEventProps: {
+						menu_item: 'A4A / Overview',
+					},
+					isSelected: true,
+				},
+				{
+					icon: category,
+					path: '/',
+					link: '/sites',
+					title: 'Sites',
```
Comment (yashwin) [inline client/a8c-for-agencies/controller.tsx:28]: Can we add translated text here?

File: client/a8c-for-agencies/controller.tsx [hunk 8]
```diff
@@ -1,9 +1,75 @@
...
+export const redirectToOverviewContext: Callback = () => {
+	page( '/overview' );
+};
+
+export const overviewContext: Callback = ( context, next ) => {
+	context.secondary = (
+		<Sidebar
+			path="/"
+			menuItems={ [
+				{
```
Comment (yashwin) [inline client/a8c-for-agencies/controller.tsx:14]: Do you think it's better to use `createItem` method to create menu items?

File: client/a8c-for-agencies/controller.tsx [hunk 9]
```diff
@@ -1,9 +1,75 @@
...
+					icon: home,
+					path: '/',
+					link: '/overview',
+					title: 'Overview',
+					trackEventProps: {
```
Comment (yashwin) [inline client/a8c-for-agencies/controller.tsx:19]: Can we also add `trackEventName` to all the menu items?

File: client/a8c-for-agencies/controller.tsx [hunk 10]
```diff
@@ -1,9 +1,75 @@
...
+						menu_item: 'A4A / Overview',
+					},
+					isSelected: true,
```
Comment (yashwin) [inline client/a8c-for-agencies/controller.tsx:22]: isSelected to set to true for this menu item. Are we planning to update this later?

File: client/a8c-for-agencies/components/sidebar/use-outside-click-callback.ts
Comment (yashwin) [inline client/a8c-for-agencies/components/sidebar/use-outside-click-callback.ts:1]: This file can be reused from `/client/lib/use-outside-click-callback/index.ts`

File: client/a8c-for-agencies/components/sidebar/style.scss [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+			color: #6f6f6f;
+		}
+		&.is-active {
+			background: var(--studio-green-0);
+			color: var(--studio-green-50);
+
+			svg {
+				color: var(--studio-green-50);
+			}
+		}
```
Comment (yashwin) [inline client/a8c-for-agencies/components/sidebar/style.scss]: [NITPICK] Do you think we can keep a constant for A4A-specific colors on this file and update it with the variable once we have the theme?

File: client/a8c-for-agencies/components/sidebar/site-selector.tsx [hunk 12]
```diff
@@ -0,0 +1,59 @@
+import { useRef } from 'react';
+import BaseSiteSelector from 'calypso/components/site-selector';
+import { useDispatch, useSelector } from 'calypso/state';
+import { setLayoutFocus } from 'calypso/state/ui/layout-focus/actions';
+import { getCurrentLayoutFocus } from 'calypso/state/ui/layout-focus/selectors';
+import useOutsideClickCallback from './use-outside-click-callback';
+
+/* NOTE: Code for this component was borrowed from calypso/my-sites/picker,
+ * with some slight modifications because we can safely assume we're using
+ * Jetpack Cloud.
```
Comment (yashwin) [inline client/a8c-for-agencies/components/sidebar/site-selector.tsx]:
```suggestion
 * Automattic for Agencies.
```

File: client/a8c-for-agencies/components/sidebar/header/style.scss
Comment (yashwin) [inline client/a8c-for-agencies/components/sidebar/header/style.scss:1]: [NITPICK] Can we use RTL-specific CSS attributes here?

File: client/a8c-for-agencies/components/sidebar/header/index.tsx [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+import a8cLogo from './a8c-logo.svg';
+import ProfileDropdown from './profile-dropdown';
+
+type Props = {
+	forceAllSitesView?: boolean;
+	simple?: boolean;
+};
+
+const AllSitesIcon = () => (
+	<img className="a4a-sidebar__all-sites-icon" src={ a8cLogo } alt="" role="presentation" />
```
Comment (yashwin) [inline client/a8c-for-agencies/components/sidebar/header/index.tsx]: Let's update the logo once this [PR](https://github.com/Automattic/wp-calypso/pull/87421) is merged.

**Review by yashwin — COMMENTED** [review]
Good job on this, @jkguidaven!

Left a few comments. Let me know your thoughts on this.

File: client/a8c-for-agencies/controller.tsx (hunk 3 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/controller.tsx:50]: I apologize for not being clear earlier. The lines of code that you see here are only temporary and are being used for demonstration purposes. I will remove them as soon as I start working on the actual menu list. Please ignore any errors in translation strings or event tracking that you may come across.

File: client/a8c-for-agencies/components/sidebar/use-outside-click-callback.ts
Comment (jkguidaven) [reply client/a8c-for-agencies/components/sidebar/use-outside-click-callback.ts:1]: Good catch. This has been fixed. 👍

File: client/a8c-for-agencies/controller.tsx (hunk 4 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/controller.tsx:61]: see my comment above.

File: client/a8c-for-agencies/controller.tsx (hunk 5 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/controller.tsx:47]: see my comment above.

File: client/a8c-for-agencies/controller.tsx (hunk 6 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/controller.tsx:18]: see my comment above.

File: client/a8c-for-agencies/controller.tsx (hunk 7 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/controller.tsx:28]: see my comment above.

File: client/a8c-for-agencies/controller.tsx (hunk 8 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/controller.tsx:14]: see my comment above.

File: client/a8c-for-agencies/controller.tsx (hunk 9 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/controller.tsx:19]: see my comment above.

File: client/a8c-for-agencies/controller.tsx (hunk 10 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/controller.tsx:22]: see my comment above.

File: client/a8c-for-agencies/components/sidebar/style.scss (hunk 11 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/components/sidebar/style.scss]: Good point. I made some changes and used some existing variables (--color-sidebar-menu-*), which we can override later in the theme once we have the colors available.

File: client/a8c-for-agencies/components/sidebar/header/style.scss
Comment (jkguidaven) [reply client/a8c-for-agencies/components/sidebar/header/style.scss:1]: Can you elaborate on which one I need to update? I am not particularly sure which RTL-specific CSS I can use.

File: client/a8c-for-agencies/components/sidebar/header/index.tsx (hunk 13 above)
Comment (jkguidaven) [reply client/a8c-for-agencies/components/sidebar/header/index.tsx]: Done 👍

**Comment by jkguidaven** [comment]
@yashwin, thanks for the review; I made some changes to address a few of your comments. Can you have another look? 🙏

File: client/a8c-for-agencies/components/sidebar/header/style.scss
Comment (yashwin) [reply client/a8c-for-agencies/components/sidebar/header/style.scss:1]: Sure. `top` & `right` mainly

**Review by yashwin — APPROVED** [review]
Thanks for addressing my comments, @jkguidaven!

LGTM :shipit:

File: client/a8c-for-agencies/components/sidebar/header/style.scss
Comment (jkguidaven) [reply client/a8c-for-agencies/components/sidebar/header/style.scss:1]: this has been addressed 👍

**Comment by simison** [comment]
Reverted due bug at landing to wp.com and seeing this first:

image

When toggle enabled at /me/account:

image

**Pull request merged**